rust/agama-server/src/questions/web.rs

```diff
@@ -0,0 +1,151 @@
+#[derive(Clone)]
+struct QuestionsClient<'a> {
+    connection: zbus::Connection,
+    questions_proxy: Questions1Proxy<'a>,
+    objects_proxy: ObjectManagerProxy<'a>,
+}
+
+impl<'a> QuestionsClient<'a> {
+    pub async fn new(dbus: zbus::Connection) -> Result<Self, zbus::Error> {
+        Ok(Self {
+            connection: dbus.clone(),
+            questions_proxy: Questions1Proxy::new(&dbus).await?,
+            objects_proxy: ObjectManagerProxy::new(&dbus).await?
+        })
+    }
+
+    pub async fn questions(self) -> Result<Vec<Question>, ServiceError> {
+        // TODO: real call to dbus
+        Ok(vec![])
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum QuestionsError {
+    #[error("Question service error: {0}")]
+    Error(#[from] ServiceError),
+}
+
+impl IntoResponse for QuestionsError {
+    fn into_response(self) -> Response {
+        let body = json!({
+            "error": self.to_string()
+        });
+        (StatusCode::BAD_REQUEST, Json(body)).into_response()
+    }
+}
+
+#[derive(Clone)]
+struct QuestionsState<'a> {
+    questions: QuestionsClient<'a>,
+}
+
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct Question {
+    generic: GenericQuestion,
+    with_password: Option<QuestionWithPassword>,
+}
+
+/// Facade of agama_lib::questions::GenericQuestion
+/// For fields details see it.
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct GenericQuestion {
+    id: u32,
+    class: String,
+    text: String,
+    options: Vec<String>,
+    default_option: String,
+    data: HashMap<String, String>
+}
+
+/// Facade of agama_lib::questions::WithPassword
+/// For fields details see it.
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct QuestionWithPassword {
+    password: String
+}
+
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct Answer {
+    generic: GenericAnswer,
+    with_password: Option<PasswordAnswer>,
+}
+
+/// Answer needed for GenericQuestion
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct GenericAnswer {
+    answer: String
+}
+
+/// Answer needed for Password specific questions.
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct PasswordAnswer {
+    password: String
+}
+/// Sets up and returns the axum service for the questions module.
+pub async fn questions_service(dbus: zbus::Connection) -> Result<Router, ServiceError> {
+    let questions = QuestionsClient::new(dbus.clone()).await?;
+    let state = QuestionsState { questions };
+    let router = Router::new()
+        .route("/questions", get(list_questions))
+        .route("/questions/:id/answer", put(answer))
+        .with_state(state);
+    Ok(router)
+}
+
+pub async fn questions_stream(dbus: zbus::Connection) -> Result<impl Stream<Item = Event>, Error> {
+    let proxy = ObjectManagerProxy::new(&dbus).await?;
+    let add_stream = proxy
+        .receive_interfaces_added()
+        .await?
+        .then(|_| async move {
```

Comment:
I do not know what your plan is here, but to simplify things a bit, you could include the question number (taken from the D-Bus path) and let the client (on the JavaScript side) retrieve the question.

Reply:
well, current API does not allow to retrieve single question. That is different to DBus.
Basically my idea about API is:
1. list all unanswered questions at `/questions`
2. answer specific question according to its id with `/questions/:id/answer`
3. get event that list of unanswered questions changed
4. and now when I am thinking about it, we need also way to provide answers file, so some API for it.

Reason why I design it this way is that common workflow is question arise, answer is provided and then another question arise. I think that having multiple unanswered question is quite rare situation, so I optimize API for that single case, but do not prevent in future to extend it e.g. to also get single question if there will be many questions.

Reply:
OK, I get the idea. About the answers file, we are not using it right now, so you can skip that part.